lnwire/accept_channel.go

```diff
@@ -145,18 +144,28 @@ func (a *AcceptChannel) Decode(r io.Reader, pver uint32) error {
 		&a.DelayedPaymentPoint,
 		&a.HtlcPoint,
 		&a.FirstCommitmentPoint,
+		&a.ExtraData,
 	)
+}
+
+// UpfrontShutdownScript is the script to which the channel funds should be
+// paid when mutually closing the channel. This field is optional, and and has
+// a length prefix, so a zero will be written if it is not set and its length
+// followed by the script will be written if it is set.
+func (a *AcceptChannel) UpfrontShutdownScript() (DeliveryAddress, error) {
+	var addr DeliveryAddress
+
+	tlvs, err := a.ExtraData.ExtractRecords(addr.NewRecord())
```

Comment:
Related to the comment above, we could have the callers use methods to "attach" these fields, which exposes less details to them, and is an alternative to setting `nil` values. I still loosely favor letting the callers set the fields themselves, and hiding everything behind `Encode/Decode` re if it's actually a TLV or not. As we evolve the messages, I imagine we'll have more and more "TLV native" fields, which can start to clutter the call sites if they need to always manually pack/encode them.

Reply:
If the caller alters the exposed field without modifying the blob, the extra data blob gets out of sync with the field. A possible solution is to make the `extraData` field private, and only create it at time of encoding, packing the set fields in the process. The problem with this however, is that fields unknown to us still must be stored in the blob (imagine reading a message with a blob having unknown fields, and then attempting to add the `UpfrontShutdown` field to it. At this point we would have to add the data to the blob when encoding it, which is currently not that easy (or something we want?). We also don't know for sure whether the field is already present in the blob.

So I think there are a few avenues we can go down:

* (In this PR currently) Making the caller TLV aware such that the whole extra data blob must be set by the caller.
* Only exposing the `UpfrontShutdownField` and not the blob, and documenting that altering this field  _will invalidate_ any data already in the blob. At encoding we discard what's already in the blob and create a new one.
* Adding `Setters/Getters` for the field. The getter will simply try to fetch it from the blob, the `Setter` will add it to the blob (if the blob is non-empty with unknown fields we must either return and error at this point, or extend our TLV code to allow adding fields to an existing stream).
* Create a constructor that takes the optional fields. This way we can make sure the fields cannot be altered, and the blob cannot change after creation.

All of these are just ways to avoid the field getting out of sync with the blob of course. If we make sure to _never alter_ the field after creation, then it doesn't matter that much, but would be nice to have it somewhat user-mistake proof .

Thoughts? cc @wpaulino

Reply:
> Only exposing the UpfrontShutdownField and not the blob, and documenting that altering this field will invalidate any data already in the blob. At encoding we discard what's already in the blob and create a new one.

I favor this option personally. I don't think we should idle long on this though, either direction can easily be reversed. I'm mostly imagining the future messages related to dynamic commitments, which will likely be _entirely_ TLV based. If we keep the current API through those messages then we'll end up with a bunch of boiler plate just to create the message.